### PR TITLE
change the argument passed to CallbackQueue.getPooled

### PR DESCRIPTION
**Pull request by gwmccull**
fix to #10100

### CODE REVIEW

**Comment by gaearon** [comment]
This code has not changed in a very long time, and doesn't seem to have changed between 15 and 16.

Could you please:

1) Add a unit test for the regression
2) Explain why this wasn't a problem in 15 despite the same code at this line 

Thanks!

**Comment by gwmccull** [comment]
@gaearon ok, I think I mostly have it worked out.  The code change that actually exposed this bug was in:
```
/src/renderers/shared/stack/reconciler/ReactUpdates.js
```

In version 15.4.1, you see this code:
```
var UPDATE_QUEUEING = {
  initialize: function() {
    this.callbackQueue.reset();
  },
  close: function() {
    this.callbackQueue.notifyAll();
  },
};

function ReactUpdatesFlushTransaction() {
  /* redacted */
  this.callbackQueue = CallbackQueue.getPooled();
  /* redacted */
}
```
The setState callbacks are getting pushed into that callbackQueue and run in that context.  I can prove it by adding changing it to `CallbackQueue.getPooled(1)` and seeing that my callback function is run with a parameter of `1`.  

Since `getPooled` has no arguments, the parameter passed to the callback function will always be `undefined`.

In version 16, all of that code is gone and the CallbackQueue is apparently being taken from component instead.  That's why when I made the change in my PR, it changed the callback function from being called with null to being called with undefined.

I've created a unit test for ReactNativeReconcileTransaction but this is kind of a hard issue to test since testing the bug I actually experienced would involve unit tests on the React Native side of things.

**Comment by gaearon** [comment]
>I've created a unit test for ReactNativeReconcileTransaction but this is kind of a hard issue to test since testing the bug I actually experienced would involve unit tests on the React Native side of things.

Can you rewrite the test to verify observable API instead of internals? In this particular case, you could write a test that schedules a `setState` callback, and then assert that it does’t get extra arguments.

To force it to follow RN code path, you can [do something like this](https://github.com/facebook/react/blob/master/src/renderers/native/__tests__/ReactNativeMount-test.js#L30-L35).

**Comment by gwmccull** [comment]
I replaced the test with one that tests setState directly with a mounted component.

the format of the test is a bit odd.  Originally, I had:
```javascript
  it('calls setState with the correct parameters', () => {
    var mockCallback = jest.fn();
    class Component extends React.Component {
      componentDidMount() {
        this.setState({}, mockCallback);
      }
      render() {
        return false;
      }
    }

    ReactNative.render(<Component />, 11);
    expect(mockCallback).toBeCalledWith(undefined);
  });
```

That test worked as expected when I ran jest on it directly.  But when I ran the `scripts/fiber/record-tests` it failed.  Somehow the mock function was being called with `[undefined]` when I ran the test directly but `[]` when I ran the test with the script.

I'm not sure what is causing the difference but this gets at the arguments with a closure rather than the jest mock functions so it should be a valid test

**Comment by gaearon** [comment]
Can you please run `yarn prettier` to prettify the code and commit the result?
Thanks!

File: src/renderers/native/__tests__/ReactNativeMount-test.js [hunk 1]
```diff
@@ -108,4 +108,20 @@ describe('ReactNative', () => {
     ReactNative.render(<Component chars={after} />, 11);
     expect(UIManager.__dumpHierarchyForJestTestsOnly()).toMatchSnapshot();
   });
+
+  it('calls setState with the correct parameters', () => {
```
Comment (gaearon) [inline src/renderers/native/__tests__/ReactNativeMount-test.js]: Let's add a similar test to `ReactTestUtils-test` for shallow renderer, and to `ReactCompositeComponent-test` for DOM renderer.

Comment (gwmccull) [inline src/renderers/native/__tests__/ReactNativeMount-test.js]: done

**Comment by gwmccull** [comment]
@gaearon I created those additional tests but to get them to pass, I had to fix another instance of that bug.  It looks like that code is related to React for the web.

I don't use React on the web so I can't easily verify if this was an issue

**Comment by gwmccull** [comment]
I ran finally got prettier to run.

`npm run prettier` didn't seem to do anything.  Had to run `npm run prettier-all`

**Comment by gwmccull** [comment]
@gaearon Do I need to do anything else on this PR?  Just want to make sure this isn't waiting on me to do something

File: src/renderers/dom/test/__tests__/ReactTestUtils-test.js [hunk 2]
```diff
@@ -844,4 +844,20 @@ describe('ReactTestUtils', () => {
       );
     });
   });
+
+  it('should call setState with the correct parameters', () => {
```
Comment (gaearon) [inline src/renderers/dom/test/__tests__/ReactTestUtils-test.js]: Let's change this to `should call setState callback with no arguments` (and others too)

File: src/renderers/dom/test/__tests__/ReactTestUtils-test.js [hunk 3]
```diff
@@ -844,4 +844,20 @@ describe('ReactTestUtils', () => {
...
+    let mockArg = 'the wrong value';
+    let mockCallback = arg => (mockArg = arg);
```
Comment (gaearon) [inline src/renderers/dom/test/__tests__/ReactTestUtils-test.js]: Maybe let's assert `arguments.length` is `0` instead of checking first argument?

**Comment by gaearon** [comment]
Thanks for your work. This mostly looks good, just a few final comments.

File: src/renderers/dom/test/__tests__/ReactTestUtils-test.js (hunk 3 above)
Comment (gwmccull) [inline src/renderers/dom/test/__tests__/ReactTestUtils-test.js]: so I tried this but `npm test` causes the length of arguments to be 1, `[ undefined ]`.  But `node ./scripts/fiber/record-tests` has the length of the arguments as 0, `[ ]`

So it seems like checking the first argument is the only way to go that I see

Comment (gaearon) [inline src/renderers/dom/test/__tests__/ReactTestUtils-test.js]: Can you see where that `undefined` comes from? I’d like to understand it.

**Comment by gwmccull** [comment]
@gaearon ok, I removed `arg` and cleaned up the Flow types for CallbackQueue.  I also fixed up the tests

**Comment by gaearon** [comment]
This looks awesome. Thank you! This will get into React Native with the next sync. Note that this code doesn’t live in `react` package so you’ll need to wait for `react-native` release including new code. I don’t know their schedule for sure so it’s a bit hard for me to say if the sync will make it into next release or not. Anyway, congrats on the contribution!

**Comment by gwmccull** [comment]
@gaearon thanks for all your feedback!

**Pull request merged**